### ISSUE TITLE
Switching 'Email as Username' alters existing custom usernames to email addresses, causing LDAP sync issues

### ISSUE THREAD

**Issue by jbuechele** [keycloak/keycloak#34560]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

Switching Keycloak's setting to "Email as Username" modifies existing usernames for users with custom usernames, replacing them with their email addresses. This unexpected behavior affects configurations where LDAP is set as the user storage. Keycloak updates the usernames in its own database, but when trying to update any attribute for a user with a modified username, it attempts to apply the username change to LDAP as well, resulting in an error.

### Version

26.0.4

### Regression

- [ ] The issue is a regression

### Expected behavior

Enabling "Email as username" should only apply to new users created after the setting is changed, while existing users should retain their usernames unchanged, especially if LDAP is configured as the user store.

### Actual behavior

When "Email as Username" is enabled:
1. Keycloak automatically changes existing usernames to email addresses. This is returned by the API:
![Screenshot 2024-11-01 at 08 57 17](https://github.com/user-attachments/assets/851def1d-b4b8-4896-9e4d-d84b3242114b)

2. Attempting to update a user attribute for a user whose username has changed will cause Keycloak to attempt to push the change to LDAP, resulting in an error.

```
2024-11-01 07:01:47,561 WARN  [org.keycloak.services.resources.admin.UserResource] (executor-thread-19) Could not update user!: org.keycloak.models.ModelException: Cannot change username if the realm is not configured to allow edit the usernames
	at org.keycloak.storage.ldap.mappers.UserAttributeLDAPStorageMapper.checkDuplicateUsername(UserAttributeLDAPStorageMapper.java:190)
	at org.keycloak.storage.ldap.mappers.UserAttributeLDAPStorageMapper$1.setUsername(UserAttributeLDAPStorageMapper.java:244)
```

### How to Reproduce?

1. Configure LDAP as the user storage provider.
2. Create users with custom usernames (non-email).
3. Enable "Email as Username" in Keycloak settings.
4. Observe that usernames in Keycloak are changed to email addresses for existing users.
5. Attempt to update a user attribute for one of these affected users.

### Anything else?

Related discussions:
[Issue #33360](https://github.com/keycloak/keycloak/issues/33360#issuecomment-2388371617)
[Issue #34167](https://github.com/keycloak/keycloak/issues/34167) @ssilvert mentioned that enabling "Email as Username" should not change existing usernames, but in our case, usernames are indeed being altered, especially affecting LDAP synchronization.

**Comment by mmoreiradj** [comment]
I'll try to reproduce it and make a PR.

**Comment by pedroigor** [comment]
The `Email as username` setting forces the username to be the email. This is how Keycloak has been working for a long time.

However, in this particular case we need to make sure that the error in the description does not happen if the `Email as username` setting is enabled and the username is actually changing to be the same as the email.

**Comment by pedroigor** [comment]
About being able to keep the username value and only set the email for new users when `Email as username` is enabled, this is an RFE.

I think it is a valid requirement and we will need to introduce yet another setting to manage this behavior. As mentioned before, the changing the username to be the email is the expected behavior.

**Comment by sguilhen** [comment]
It might be worth to consider adding a setting to th LDAP mapper so that changes to particular attributes are not synced back to LDAP even if the overall config for the provider is writable.

**Comment by rasos** [comment]
We have understood the `Email as username` setting is using the e-mail as username only when a new user is created. As I assume that both fields are kept in the Keycloak database separately, there should never be an automatic sync or override of the username field, e.g. not even if the user is updated and no matter if there is an LDAP backend or not. 

For historic reasons, we have at one larger Keycloak instance both kind of usernames, as e-mail and plain text. We would face problems at some connected clients if Keycloak would be overriding usernames with the users' email. However, we would like that future registrations only need to enter an e-mail and thus would like to activate `Email as username`.

To sum up: `Email as username`  should ONLY perform in the registration process copying the `e-mail` to the `username` field and not do any background replacement actions of the username.

**Comment by pedroigor** [comment]
@rasos I also see the same. It would be a breaking change though. AFAIK, this is how it has been working for a long time. 

Not saying it is correct ...

Do you have `Edit username` also enabled?

**Comment by rasos** [comment]
We had introduced the usage of the  attribute `nickname` lately (mainly for fediverse clients such as PeerTube, where users get a unique publicly visible fediverse ID). We ask for it, when I want to authorize at a client who has it in the client scope. That way we reduce the number of mandatory registration fields. But as we wanted to "switch off" the need for entering a username with the `Email as username` setting I have seen the same effect as described above by @jbuechele . So we had to roll back. No, we do not allow that users edit their username, as this would confuse some clients, especially those connected directly to LDAP.

**Comment by pedroigor** [comment]
Sent https://github.com/keycloak/keycloak/pull/34961.

**Comment by chlunde** [comment]
```
                if (realm.isRegistrationEmailAsUsername() && username.equals(user.getEmail())) {
                    return;
                }
```

I wonder if this could have caused issues for people with `Upper.Case@example.com` emails, as other places in keycloak the username is lower cased automatically. We see that they get "Update Account Information" prompts now.
